File: tools/AtlasBuildTool/CMake/CMakeTypes.cs
// Copyright(c) 2023-present, Atlas.
// Distributed under the MIT License (http://opensource.org/licenses/MIT)

namespace AtlasBuildTool.CMake;

public static class CMakeTypes
{
    public const string EditorOnly = "EDITOR_ONLY";
    public const string SkipMoc = "SKIP_MOC";
    public const string TestTarget = "TEST_TARGET";
    public const string GameTarget = "GAME_TARGET";
    public const string Target = "TARGET";
    public const string PublicDefinitions = "PUBLIC_DEFINITIONS";
    public const string PrivateDefinitions = "PRIVATE_DEFINITIONS";
    public const string PublicIncludeDirs = "PUBLIC_INCLUDE_DIRS";
    public const string PrivateIncludeDirs = "PRIVATE_INCLUDE_DIRS";
    public const string PublicLinkLib = "PUBLIC_LINK_LIB";
    public const string PrivateLinkLib = "PRIVATE_LINK_LIB";
    public const string DependencyTargets = "DEPENDENCY_TARGETS";
    public const string PlatformCodeDirs = "PLATFORM_CODE_DIRS";

    public static readonly string[] Options = { EditorOnly, SkipMoc, TestTarget, GameTarget };
    public static readonly string[] OneValueArgs = { Target };
    public static readonly string[] MultiValueArgs =
    {
        PublicDefinitions, PrivateDefinitions, PublicIncludeDirs, PrivateIncludeDirs,
        PublicLinkLib, PrivateLinkLib, DependencyTargets, PlatformCodeDirs
    };

    public const string AddLibraryFunction = "add_atlas_library";
    public const string AddExecutableFunction = "add_atlas_executable";
}

File: tools/atlas_builder/src/cmake/CMakeTypes.kt
/*
 * Copyright(c) 2023-present, Atlas.
 * Distributed under the MIT License (http://opensource.org/licenses/MIT)
 */

package com.spacelion.atlas_builder.cmake

class CMakeTypes private constructor() {
    companion object {
        val EDITOR_ONLY = "EDITOR_ONLY"
        val TEST_TARGET = "TEST_TARGET"
        val GAME_TARGET = "GAME_TARGET"
        val TARGET = "TARGET"
        val PUBLIC_DEFINITIONS = "PUBLIC_DEFINITIONS"
        val PRIVATE_DEFINITIONS = "PRIVATE_DEFINITIONS"
        val PUBLIC_INCLUDE_DIRS = "PUBLIC_INCLUDE_DIRS"
        val PRIVATE_INCLUDE_DIRS = "PRIVATE_INCLUDE_DIRS"
        val PUBLIC_LINK_LIB = "PUBLIC_LINK_LIB"
        val PRIVATE_LINK_LIB = "PRIVATE_LINK_LIB"
        val DEPENDENCY_TARGETS = "DEPENDENCY_TARGETS"
        val PLATFORM_CODE_DIRS = "PLATFORM_CODE_DIRS"

        val OPTIONS = setOf(EDITOR_ONLY, TEST_TARGET, GAME_TARGET)
        val ONE_VALUE_ARGS = setOf(TARGET)
        val MULTI_VALUE_ARGS = setOf(PUBLIC_DEFINITIONS, PRIVATE_DEFINITIONS, PUBLIC_INCLUDE_DIRS,
            PRIVATE_INCLUDE_DIRS, PUBLIC_LINK_LIB, PRIVATE_LINK_LIB, DEPENDENCY_TARGETS, PLATFORM_CODE_DIRS)

        val ADD_LIBRARY_FUNCTION = "add_atlas_library"
        val ADD_EXECUTABLE_FUNCTION = "add_atlas_executable"
        val ADD_PROJECT_FUNCTION = "add_atlas_project"
    }
}